vulnerable_people_form/integrations/postcode_lookup_helper.py

```diff
@@ -30,50 +30,102 @@ def address_line_builder(lpi_info, slice):
+    primary_property_identifier_is_a_building = True if lpi_info.get('PAO_TEXT') else False
+
+    secondary_property_identifier = property_identifier(start_number=lpi_info.get('SAO_START_NUMBER'),
+                                                        start_suffix=lpi_info.get('SAO_START_SUFFIX'),
+                                                        end_number=lpi_info.get('SAO_END_NUMBER'),
+                                                        end_suffix=lpi_info.get('SAO_END_SUFFIX'),
+                                                        text_value=lpi_info.get('SAO_TEXT'))
+
     street = lpi_info.get("STREET_DESCRIPTION", "")
+    organisation = lpi_info.get("ORGANISATION", "")
 
-    if sao != "":
-        building_and_street_line_1 = sao
-        building_and_street_line_2 = pao + " " + street
-    elif lpi_info.get("PAO_TEXT"):
-        building_and_street_line_1 = pao
-        building_and_street_line_2 = street
-    else:
-        building_and_street_line_1 = pao + " " + street
+    building_and_street_line_1, building_and_street_line_2 = build_building_and_street_lines(
+        primary_property_identifier, primary_property_identifier_is_a_building, secondary_property_identifier, street)
+
+    if organisation:
+        building_and_street_line_1, building_and_street_line_2 = include_organisation_name(building_and_street_line_1,
+                                                                                           building_and_street_line_2,
+                                                                                           organisation)
 
     return {
-        "uprn": int(lpi_info.get("UPRN")),
-        "town_city": address_line_builder(lpi_info, ["LOCALITY_NAME", "TOWN_NAME", "ADMINISTRATIVE_AREA"]).title(),
+        "uprn": int(lpi_info.get("UPRN")) if lpi_info.get("UPRN") else None,
+        "town_city": town_city_builder(lpi_info).title(),
         "postcode": lpi_info.get("POSTCODE_LOCATOR"),
         "building_and_street_line_1": building_and_street_line_1.title(),
         "building_and_street_line_2": building_and_street_line_2.title(),
     }
 
 
+def town_city_builder(lpi_info):
+    locality = lpi_info.get('LOCALITY_NAME')
+    town = lpi_info.get('TOWN_NAME')
+    administrative_area = lpi_info.get('ADMINISTRATIVE_AREA')
+
+    non_null_sections = [item for item in [locality, town, administrative_area] if item]
+
+    unique_sections = unique_values_preserving_order(non_null_sections)
+
+    return ', '.join(unique_sections)
+
+
+def unique_values_preserving_order(sections):
+    unique_sections = []
+    for section in sections:
+        if section not in unique_sections:
+            unique_sections.append(section)
+    return unique_sections
+
+
+def property_identifier(start_number, start_suffix, end_number, end_suffix, text_value):
+    def number_suffix(a, b):
+        return f'{a}{b}' if b else f'{a}'
+
+    if text_value:
+        return text_value
```

Comment:
Quick question. If we have PAO_TEXT, should we discard the rest of the information e.g. house number etc. This is different from the current implementation.

Reply:
I thought PAO_TEXT was only used as a last resort so would never have numbers but apparently there are cases which need to be arranged with PAO_TEXT first - I'll get on it
```
      "UPRN" : "10003191910",
      "ADDRESS" : "LONDON HOUSE, 25, HIGH STREET, LINTON, CAMBRIDGESHIRE, CB21 4HS",
      "USRN" : "35300089",
      "LPI_KEY" : "0530L000009065",
      "PAO_START_NUMBER" : "25",
      "PAO_TEXT" : "LONDON HOUSE",
      "STREET_DESCRIPTION" : "HIGH STREET",
      "TOWN_NAME" : "LINTON",
      "ADMINISTRATIVE_AREA" : "CAMBRIDGESHIRE",
      
```